addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/discovery/RioSystemDeviceDiscoveryService.java

```diff
@@ -0,0 +1,270 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.russound.internal.discovery;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.russound.internal.RussoundHandlerFactory;
+import org.openhab.binding.russound.internal.net.SocketChannelSession;
+import org.openhab.binding.russound.internal.net.SocketSession;
+import org.openhab.binding.russound.internal.net.WaitingSessionListener;
+import org.openhab.binding.russound.internal.rio.RioConstants;
+import org.openhab.binding.russound.internal.rio.controller.RioControllerConfig;
+import org.openhab.binding.russound.internal.rio.source.RioSourceConfig;
+import org.openhab.binding.russound.internal.rio.system.RioSystemHandler;
+import org.openhab.binding.russound.internal.rio.zone.RioZoneConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This implementation of {@link DiscoveryService} will scan a RIO device for all controllers, source and zones attached
+ * to it.
+ *
+ * @author Tim Roberts
+ *
+ */
+public class RioSystemDeviceDiscoveryService extends AbstractDiscoveryService {
+    /** The logger */
+    private final static Logger logger = LoggerFactory.getLogger(RioSystemDeviceDiscoveryService.class);
+
+    /** The system handler to scan */
+    private final RioSystemHandler sysHandler;
+
+    /** Pattern to identify controller notifications */
+    private final Pattern RSP_CONTROLLERNOTIFICATION = Pattern.compile("(?i)^[SN] C\\[(\\d+)\\]\\.(\\w+)=\"(.*)\"$");
```

Comment:
Pretty sure I got them all (over about 7 classes).  Let me know if I didn't on the next commit